generate_small_test_ref.py

```diff
@@ -16,96 +15,132 @@ def uint(value):
+    pysam.faidx(out_reference)
+    logger.info("Lifted over the reference to {}".format(out_reference))
+
+    reference_handle.close()
+
+    return out_reference
+
+
+def gen_restricted_vcf(in_vcf, regions_bed, out_vcf, restricted_reference, targeted_samples, flank=0, use_short_contig_names=False):
+    logger = logging.getLogger(gen_restricted_vcf.__name__)
+
+    if not in_vcf:
+        return None
+
+    if not os.path.isfile(in_vcf):
+        logger.error("%s not found" % in_vcf)
+        return None
+
+    reference_handle = pysam.Fastafile(restricted_reference)
+    contigs = list(zip(reference_handle.references, reference_handle.lengths))
+    reference_handle.close()
+
+    # get the base name and use it in the output
+    vcf_template_reader = vcf.Reader(open(in_vcf, "r"))
+    vcf_template_reader.metadata["reference"] = restricted_reference
+    vcf_template_reader.contigs = OrderedDict([(contig_name, vcf.parser._Contig(contig_name, contig_length)) for (contig_name, contig_length) in contigs])
+
+    new_samples = []
+    if targeted_samples:
+        for k,v in sorted(vcf_template_reader._sample_indexes.iteritems()):
+            if k in targeted_samples:
+                new_samples.append(k)
+        vcf_template_reader.samples = new_samples
+
+    vcf_writer = vcf.Writer(open(out_vcf, "w"), vcf_template_reader)
+
+    if targeted_samples:
+        vcf_template_reader = vcf.Reader(open(in_vcf, "r"))
+
+    #tabix_vcf = pysam.TabixFile(invcf, parser=pysam.asVCF())
+    info_warned = False
+    regions_bedtool = pybedtools.BedTool(regions_bed)
+
+    for region_index, region in enumerate(regions_bedtool, start=1):
+        records = None
+        try: records = vcf_template_reader.fetch(chrom=str(region.chrom), start=region.start, end=region.end)
+        except ValueError: logger.error("Failed to retrieve %s from %s" % (str(region).strip(), in_vcf))
+        if records is None: continue
+        for record in records:
+            if record.POS <= region.start + flank or record.POS + len(record.REF) + flank - 1 >= region.end: continue
+            record.CHROM = str(region_index) if use_short_contig_names else ("%s_%d_%d" % (str(region.chrom), region.start, region.end))
+            # record.POS seems to be zero-based, at least in the infinite wisdom of my version of pysam
+            record.POS = record.POS - region.start
+            if not new_samples:
+                vcf_writer.write_record(record)
+                continue
+            else:
+                snames = []
+                sindexes = {}
+            for s in new_samples:
+                for i in xrange(len(record.samples)):
+                    if s == record.samples[i].sample:
+                        sindexes[s] = i
+                        snames.append(record.samples[i])
+            vcfrecord = vcf.model._Record(record.CHROM, record.POS, record.ID, record.REF, record.ALT, record.QUAL, record.FILTER, record.INFO, record.FORMAT, sindexes, snames)
```

Comment:
Hi @marghoob, I got an exception that seems to be caused by this line: ```python /net/hippo/volumes/wadi/shared/prj/guoy28/Downloads/allamerican_oncology_build/examples/brca_uid_linker/../..//opt//varsim-git/varsim_multi.py --out_dir /net/hippo/volumes/wadi/shared/prj/guoy28/Downloads/allamerican_oncology_build/examples/brca_uid_linker/simulation_out_100x_len500_hereditary_onco_uid_linker --reference /net/kodiak/volumes/delta/shared/home/rogerliu/databases/hg/hs37d5.fa --samples NA12878 --simulator longislnd --vcfs /net/hippo/volumes/wadi/shared/prj/genia/applications/allamerican-test/1kgvcfs/ALL.combined.hg19.vcf.gz --regions /net/hippo/volumes/wadi/shared/prj/guoy28/Downloads/allamerican_oncology_build/examples/brca_uid_linker/filled500bp.bed --disable_sim --sampling_vcf BRCA12.pathogenic.hs37.vcf.gz --disable_rand_dgv --lift_ref --disable_rand_dgv --seed 11 --vc_num_snp 25 --vc_num_ins 1 --vc_num_del 1
Traceback (most recent call last):
  File "/net/hippo/volumes/wadi/shared/prj/guoy28/Downloads/allamerican_oncology_build/examples/brca_uid_linker/../..//opt//varsim-git/varsim_multi.py", line 214, in <module>
    disable_sim=args.disable_sim)
  File "/net/hippo/volumes/wadi/shared/prj/guoy28/Downloads/allamerican_oncology_build/examples/brca_uid_linker/../..//opt//varsim-git/varsim_multi.py", line 54, in varsim_multi
    restricted_reference, restricted_vcfs = gen_restricted_ref_and_vcfs(reference, variant_vcfs + [sampling_vcf], regions, samples, restricted_dir, flank=0, short_contig_names=False)
  File "/net/hippo/volumes/wadi/shared/prj/guoy28/Downloads/allamerican_oncology_build/opt/varsim-git/generate_small_test_ref.py", line 118, in gen_restricted_ref_and_vcfs
    generated_vcfs.append(gen_restricted_vcf(invcf, regions, outvcf, restricted_fasta, samples, flank, short_contig_names))
  File "/net/hippo/volumes/wadi/shared/prj/guoy28/Downloads/allamerican_oncology_build/opt/varsim-git/generate_small_test_ref.py", line 94, in gen_restricted_vcf
    vcfrecord = vcf.model._Record(record.CHROM, record.POS, record.ID, record.REF, record.ALT, record.QUAL, record.FILTER, record.INFO, record.FORMAT, sindexes, snames)```

the issue can be eliminated by adding declaration at line 83 for `snames` and `sindexes`.